### ISSUE TITLE
Question: vocoder / gan application, bf16

### ISSUE THREAD

**Issue by codename0og** [muooon/EmoNavi#1]
Hi!
Do you happen to know if it's worth trying ti emonavi on gan vocoders? And if so, how does it behave in bf16?

Also, currently I use a bf16 variant of AdamW ( tf32, bf16 amp ) which is really amazing in stability + has error correction / kahan / compensation ~

I believe you could perhaps try to add something like that in for bf16 purposes?
https://github.com/lessw2020/bfloat_optimizer/blob/main/bfloat_optimizer.py

In any case, looking forward to your reply!

**Comment by muooon** [comment]
[codename0og](https://github.com/codename0og)

Thank you very much for your interest in this optimizer.

I haven't used "GAN vocoders" myself, so I can't speak to how it performs in that context. However, I believe there are advantages that make it worth trying. Since it is AMP-compatible, bf16 should work without any issues. (Mixed precision bf16 has been confirmed to work during LoRA creation for SDXL.)

As for "error compensation techniques such as Kahan summation," they are not implemented within emonavi. If such functionality is required, unfortunately, this optimizer does not support it. If the tool you are using provides such features, they may be usable—but I recommend checking with the tool's developer to be sure.

—Benefits—  
GAN-based vocoders tend to suffer from gradient instability >> emonavi helps suppress overfitting and divergence

Thank you again for your question. Feel free to reach out anytime if you have further inquiries.

—P.S.—  
The emotional mechanism has been updated from emonavi (first generation) >> to emosens (second generation).  
If you're interested, please try the emosens generation.

この optimizer に、興味を持って頂きありがとうございます。
"gan vocoders" を使用したことがないため、使用感はわかりませんが、利用すべき"利点"はあると思います。 また AMP対応済みなので、bf16 も問題なく動作すると思います。(SDXL の LoRA 作成では混合精度 bf16 でも動作確認済み)
そして"Kahan加算などの誤差補償技術"は、emonavi内部にはありません。これを必要な場合には、残念ながら、この optimizer では対応できない、と思います。ツール側にある機能であれば利用可能かもしれませんが、それはツール側の作成者等にお尋ね頂いたほうが確実だろうと思います。
－利点－
GANベースのVocoderは勾配が不安定になりやすい >> emonavi は過学習や発散を抑制します
　
このたびはご質問を頂きありがとうございました。また何かあればご投稿ください。
　
－追伸－
感情機構について emonavi (第一世代) >> emosens (第二世代) へ update しました。
もし気になりましたら emosens 世代をお使いください。

**Comment by codename0og** [comment]
Thank you for reply! yes, I will surely give the emosense a shot.
Have a nice day!

**Comment by muooon** [comment]
> [codename0og](https://github.com/codename0og)
> Thank you for reply! yes, I will surely give the emosense a shot. Have a nice day!

Hello. It's been about two months since you asked your question. How have things been since then? I looked into Kahan compensation. What I realized is that the current EmoNAVI and its variation (v3.0) incorporate a mechanism that maintains accuracy without relying on Kahan compensation. I believe this emo-based system will continue to update while preserving sufficient accuracy during voice AI learning and training. Furthermore, since it refines without requiring additional tensors like Kahan compensation, we expect it to enable lightweight yet accurate updates. Please do try (v3.0).

こんにちは。ご質問を頂き２か月ほどを経過しましたが、その後いかがでしょうか。kahan補償 について調べました。それで気づいたのですが、現 EmoNAVI とバリエーション(v3.0)は、kahan補償 とは違いますが、精度を落とさない仕組みを備えています。音声AIの学習や訓練で、この emo系 は十分な精度を維持し更新を行うものと思います。また kahan補償 のような追加テンソルを持たずに精緻化していますから、軽量かつ正確な更新を行えると期待できると思います。ぜひ (v3.0) をお試しくださいね。

**Comment by codename0og** [comment]
Can give it a try soon, atm I am finalizing my vocoder experiments and once that's done, gonna give v3 a go

**Comment by muooon** [comment]
> [codename0og](https://github.com/codename0og)

How has it been going since then? I have no experience training speech models, but in SDXL training, experiments showed that using low precision (fp8) only resulted in errors equivalent to fp16. This optimizer was found to be effective (since the standard learning rate works well at lower values, try reducing it to about half or one-fifth of the Adam case). I expect the same will hold for speech models. If you're training at high precision, why not test at a lower precision? You might achieve comparable results faster than with other optimizers. Reducing precision for speed could potentially save training time.

その後はいかがでしょうか。音声モデルの学習は未経験でわかりませんが、SDXLの学習では、fp8 の低精度で、fp16 相当の誤差しか発生しない、実験でそういうoptimizerであることがわかりました (通常の学習率は低めで機能しますから、Adamのケースの1/2程度、または1/5程度まで低値にしてみてください) それで音声モデルでも同様だろうと予想しています。もし高精度での学習をしているのであれば、それより低精度でテストしてみてはいかがでしょう、他のoptimizerより高速で同等の結果を得られるかもしれません。軽量化で高速化になれば学習時間等を節約できる可能性があります。

**Comment by codename0og** [comment]
We've determined that bf16 most likely isn't sufficient enough for voice-synthesis,
and so, the decision was to keep it either fp32/tf32 or amp fp16..
Unlike LLMs, vision and generally image generation, audio machine learning benefits from sufficient precision it seems.

Now, if we speak purely of EmoNavi I have 2 specific issues:

1. I've noticed that the EmoNavi depends on closure which isn't how we handle it in our codebase.
We do it manually:
```
            # Discriminator backward and update:
            optim_d.zero_grad(set_to_none=True)
            if train_dtype == torch.float16:
                gradscaler.scale(loss_disc).backward() # Scale and backward of the loss
                gradscaler.unscale_(optim_d) # Unscale
                grad_norm_d = torch.nn.utils.clip_grad_norm_(net_d.parameters(), max_norm=999999) # Grad clipping
                grad_norm_d_clipped = commons.get_total_norm([p.grad for p in net_d.parameters() if p.grad is not None], norm_type=2.0, error_if_nonfinite=False) # Grad retrieval for logging
                gradscaler.step(optim_d) # Optim step
            else:
                loss_disc.backward() # Loss backward
                grad_norm_d = torch.nn.utils.clip_grad_norm_(net_d.parameters(), max_norm=999999) # Grad clipping
                grad_norm_d_clipped = commons.get_total_norm([p.grad for p in net_d.parameters() if p.grad is not None], norm_type=2.0, error_if_nonfinite=True) # Grad retrieval for logging
                optim_d.step() # Optim step

....

            # Generator backward and update:
            optim_g.zero_grad(set_to_none=True)
            if train_dtype == torch.float16:
                gradscaler.scale(loss_gen_total).backward() # Scale and backward of the loss
                gradscaler.unscale_(optim_g) # Unscale
                grad_norm_g = torch.nn.utils.clip_grad_norm_(net_g.parameters(), max_norm=999999) # Grad clipping
                grad_norm_g_clipped = commons.get_total_norm([p.grad for p in net_g.parameters() if p.grad is not None], norm_type=2.0, error_if_nonfinite=False) # Grad retrieval for logging
                gradscaler.step(optim_g) # Optim step
                gradscaler.update() # Scaler update, to prepare the scaling for the next iteration
            else:
                loss_gen_total.backward() # Loss backward
                grad_norm_g = torch.nn.utils.clip_grad_norm_(net_g.parameters(), max_norm=999999) # Grad clipping
                grad_norm_g_clipped = commons.get_total_norm([p.grad for p in net_g.parameters() if p.grad is not None], norm_type=2.0, error_if_nonfinite=True) # Grad retrieval for logging
                optim_g.step() # Optim step

```

2. FP16 training involves usage of gradscaler and I ain't quite sure how to make it work right

If you have any ideas or insights, please let me know ~
Here's the [training script](https://github.com/codename0og/codename-rvc-fork-4/blob/main/rvc/train/train.py) in case you wanna take a look ~

**Comment by muooon** [comment]
> [codename0og](https://github.com/codename0og)

This is quite a challenge. I tried modifying the emonavi code, but releasing it could potentially compromise PyTorch's standard, so I decided against publishing the modified code. I'm hoping for RVC's evolution to utilize closures. Closures are an important standard feature in various scenarios, including AMP.
And thank you for the thorough explanation regarding audio learning and fp32, bf16, fp16. Audio requires very fine-grained expression, so precision is crucial, right? For emo-style optimizers, stability is the top priority. I believe that by advancing closure support and letting users feel this stability-first approach in updates, they'll come to appreciate it.

これは大変ですね、emonavi のコード修正も試しましたが、それを公開すると PyTorch の標準を棄損する可能性があります、ゆえに修正コードの公開は諦めました。closure をつかう、RVC側の進化を期待します。AMP含むさまざまな場面で closure は重要な標準機能ですから。
そして 音声学習と fp32、bf16、fp16、について、丁寧な説明をありがとうございました。音声はとても微細な表現を必要とするでしょうから精度は大切ですよね。emo系のoptimizerは、安定最優先です。closure 対応を進めて頂いて、安定最優先の更新を感じて頂けたら気に入ってもらえるのではないか、そう思っています。

**Comment by muooon** [comment]
Initially, I attempted to modify the emonavi side. Considering future implications, I concluded it would be better not to release the emonavi fix.
What follows may be considered meddling. Regarding the tarain.py you provided, consulting Copilot suggests that the following simple modification should make it closure-compatible:

最初に emonavi 側の修正を試みました、それで今後のことも考えると、emonaviの修正は公開しない方が良い、だろうと判断しました。
ここからは余計なお世話になるかもしれません。ご提示の tarain.py について、Copilot に聞いてみると、以下のような修正だけで、closure 対応可となるようです。

```
def closure():
    optim_g.zero_grad(set_to_none=True)
    gradscaler.scale(loss_gen_total).backward()
    gradscaler.unscale_(optim_g)
    clip_grad_norm_(...)
    return loss_gen_total

gradscaler.step(optim_g, closure=closure)
gradscaler.update()
```

**Comment by codename0og** [comment]
Hmmm.. To be honest, I generally try to avoid deviating from my estabilished code's structure but, we'll see.
Gonna do some test regarding that later, if I succeed, gonna report back ~

**Comment by muooon** [comment]
> [codename0og](https://github.com/codename0og)

Modifying long code is quite a hassle, isn't it? However, once you've implemented closure support, you'll be able to use adapters like dadapt and Prodigy. You'll also gain access to options that utilize confidence metrics, such as AdaBelief, TrustRatio, and Lookahead. Furthermore, if you can support Pytorch-Optimizer, you'll gain access to numerous additional options. And with closure support, you'll also be able to use TensorBoard. By utilizing standard features, you should be able to gain access to several tools that can aid your training process. Please consider it.

長いコードを改修するのは大変ですよね。しかし closure 対応を済ませれば、dadapt系、Prodigy等も使えるようになりますし、AdaBelief、TrustRatio系、Lookahead系などの 信頼度 をつかうものも選択肢になります。また Pytorch-Optimizer にも対応できれば数多くの選択肢を得られます。そして closure 対応により、TensorBoard も使えるようになります。標準機能を利用することで学習の助けになるものをいくつも得られるはずです。ぜひご検討くださいね。
